### PR TITLE
fix: limit the amount of pending-accept reset streams

### PR DESCRIPTION
**Pull request by seanmonstar**
Streams that have been received by the peer, but not accepted by the user, can also receive a RST_STREAM. This is a legitimate pattern: one could send a request and then shortly after, realize it is not needed, sending a CANCEL.

However, since those streams are now "closed", they don't count towards the max concurrent streams. So, they will sit in the accept queue, using memory.

In most cases, the user is calling `accept` in a loop, and they can accept requests that have been reset fast enough that this isn't an issue in practice.

But if the peer is able to flood the network faster than the server accept loop can run (simply accepting, not processing requests; that tends to happen in a separate task), the memory could grow.

So, this introduces a maximum count for streams in the pending-accept but remotely-reset state. If the maximum is reached, a GOAWAY frame with the error code of ENHANCE_YOUR_CALM is sent, and the connection marks itself as errored.

Closes https://github.com/hyperium/hyper/issues/2877 (we assume)

### CODE REVIEW

**Comment by seanmonstar** [comment]
I started to make the limit configurable, but then paused, wanting to get something working faster. If it seems worth the delay, I can add in the public API to set these limits yourself.

**Comment by LPardue** [comment]
Since this seems like something that occurs due to several dynamic factors or behaviours, a single hard coded limit _on paper_ seems too restrictive. Not least because it isn't immediately obvious where the value `20` comes from. This might be an entirely reasonable value but without some wider testing, who knows? A default value that can be tweaked via API seems like a decent middleground to me.

**Review by Noah-Kennedy — APPROVED** [review]
I'm in favor of merging this now as is, but we should in the future make the threshold configurable.

**Comment by seanmonstar** [comment]
Added another commit making the options part of the public API.

File: src/client.rs [hunk 1]
```diff
@@ -966,6 +971,37 @@ impl Builder {
         self
     }
 
+    /// Sets the maximum number of pending-accept remotely-reset streams.
+    ///
+    /// When a
```
[the remark this answers is not in the document]
Comment (seanmonstar) [reply src/client.rs:976]: oh woops, I was supposed to copy the explanation from the server builder! 😴

**Comment by Noah-Kennedy** [comment]
I'm summarizing a conversation @seanmonstar and I had in Discord here:

There is a non-zero chance that the threshold we set does trip on real (and non-malicious) traffic. We have a log line in here for if this goes off, but it might be worthwhile for us to expose an actual metric for these GOAWAYs so that users of the library can better understand how frequently this trips.

**Comment by seanmonstar** [comment]
Given that I've been at this for 14 hours, I'm going to go knock out. Since the assumed problem is quite hard to trigger, and that the fix here is a guess, and that there's always some amount of risk when publishing new versions, I'm opting for leaving this open for the night.

**Comment by Noah-Kennedy** [comment]
This will also give our peers outside of North America a chance to look at this change before we release it.

If anyone has any concerns about this change, please make them heard.

**Review by edevil — APPROVED** [review]

File: src/proto/streams/recv.rs [hunk 2]
```diff
@@ -741,12 +741,39 @@ impl Recv {
...
+        // See https://github.com/hyperium/hyper/issues/2877
+        if stream.is_pending_accept {
+            if counts.can_inc_num_remote_reset_streams() {
+                counts.inc_num_remote_reset_streams();
+            } else {
+                tracing::warn!(
+                    "recv_reset; remotely-reset pending-accept streams reached limit ({:?})",
+                    counts.max_remote_reset_streams(),
+                );
+                return Err(Error::library_go_away(Reason::ENHANCE_YOUR_CALM));
```
Comment (nox) [inline src/proto/streams/recv.rs:766]: Can you drive-by expose `Error::is_library` or something like that, so we can check for `error.is_library() && error.reason() == Some(Reason::ENHANCE_YOUR_CALM)` for logging purposes?

Comment (seanmonstar) [reply src/proto/streams/recv.rs:766]: Would checking `!error.is_remote() && error.is_go_away()` be fine?

Comment (nox) [reply src/proto/streams/recv.rs:766]: There are 3 error initiators: remote, user and library, and there is no way to check for the user one either. If you don't want to expose `is_library` I'm fine with checking `!error.is_remote() && !error.is_user() && error.is_go_away() && error.reason() == Some(Reason::ENHANCE_YOUR_CALM)`

Comment (seanmonstar) [reply src/proto/streams/recv.rs:766]: We _can_ add the method. I was just wondering if what exists is enough.

As an alternative (or in addition to), we could add some debug data to GOAWAY frame, maybe `too_many_resets` or something. Oh, just looking, a method doesn't exist to access that either...

Comment (seanmonstar) [reply src/proto/streams/recv.rs:766]: Made in a separate PR: https://github.com/hyperium/h2/pull/669
